src/Parser/LRUCache.php

```diff
@@ -0,0 +1,77 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Parser;
+
+use LogicException;
+use SplDoublyLinkedList;
+use function count;
+use function sprintf;
+
+/**
+ * @template TValue
+ * @implements ParserCache<TValue>
+ */
+class LRUCache implements ParserCache
+{
+
+	/** @var array<string, TValue> */
+	private array $cache;
+
+	/** @var SplDoublyLinkedList<string> */
+	private SplDoublyLinkedList $list;
+
+	public function __construct(private int $capacity)
+	{
+		$this->cache = [];
+		$this->list = new SplDoublyLinkedList();
+	}
+
+	/**
+	 * @return TValue
+	 */
+	public function get(string $key)
+	{
+		if (!isset($this->cache[$key])) {
+			throw new LogicException(sprintf('Key %s was not found in the cache, use ->has() first', $key));
+		}
+
+		$this->list->rewind();
+		while ($this->list->current() !== $key) {
+			$this->list->next();
+		}
+		$this->list->rewind();
```

Comment:
I wonder why you rewind after the loop?

Reply:
No idea how this ended up here, definitely doesnt make sense :D